plugins/out_pgsql/pgsql_connections.c

```diff
@@ -0,0 +1,184 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "pgsql.h"
+
+void pgsql_destroy_connections(struct flb_pgsql_config *ctx)
+{
+    struct mk_list *tmp;
+    struct mk_list *head;
+    struct flb_pgsql_conn *conn;
+    PGresult *res = NULL;
+
+    mk_list_foreach_safe(head, tmp, &ctx->conn_queue) {
+        conn = mk_list_entry(head, struct flb_pgsql_conn, _head);
+        if(PQstatus(conn->conn) == CONNECTION_OK) {
+            while(PQconsumeInput(conn->conn) == 0) {
+                res = PQgetResult(conn->conn);
+                if(PQresultStatus(res) != PGRES_COMMAND_OK) {
+                    flb_plg_warn(ctx->ins, "%s",
+                                 PQerrorMessage(conn->conn));
+                }
+                PQclear(res);
+            }
+        }
+        PQfinish(conn->conn);
+        flb_free(conn);
+    }
+}
+
+void *pgsql_create_connection(struct flb_pgsql_config *ctx)
+{
+    struct flb_pgsql_conn *conn;
+
+    conn = flb_calloc(1, sizeof(struct flb_pgsql_conn));
+    conn->conn = PQsetdbLogin(ctx->db_hostname,
+                              ctx->db_port,
+                              NULL, NULL,
+                              ctx->db_name,
+                              ctx->db_user,
+                              ctx->db_passwd);
+
+    if(PQstatus(conn->conn) != CONNECTION_OK) {
+        flb_plg_error(ctx->ins,
+                      "failed connecting to host=%s with error: %s",
+                      ctx->db_hostname, PQerrorMessage(conn->conn));
+        PQfinish(conn->conn);
+        flb_free(conn);
+        return NULL;
+    }
+
+    flb_plg_info(ctx->ins, "switching postgresql connection "
+                 "to non-blocking mode");
+
+    if(PQsetnonblocking(conn->conn, 1) != 0) {
+        flb_plg_error(ctx->ins, "non-blocking mode not set");
+        PQfinish(conn->conn);
+        flb_free(conn);
+        return NULL;
+    }
+
+    return conn;
+}
+
+int pgsql_start_connections(struct flb_pgsql_config *ctx)
+{
+    int i;
+    struct flb_pgsql_conn *conn = NULL;
+
+    mk_list_init(&ctx->conn_queue);
+    ctx->active_conn = 0;
+
+    for(i = 1; i <= ctx->min_pool_size; i++) {
+        flb_plg_info(ctx->ins, "Opening connection: #%d", i);
+
+        conn = (struct flb_pgsql_conn *)pgsql_create_connection(ctx);
+        if(conn == NULL) {
+            pgsql_conf_destroy(ctx);
+            return -1;
+        }
+
+        conn->number = i;
+        ctx->active_conn++;
+        mk_list_add(&conn->_head, &ctx->conn_queue);
+    }
+
+    ctx->conn_current = mk_list_entry_last(&ctx->conn_queue,
+                                           struct flb_pgsql_conn,
+                                           _head);
+
+    return 0;
+}
+
+int pgsql_new_connection(struct flb_pgsql_config *ctx)
+{
+    struct flb_pgsql_conn *conn = NULL;
+
+    if(ctx->active_conn >= ctx->max_pool_size)
```

Comment:
use braces: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions